src/Storage/StorageObject.php

```diff
@@ -633,6 +636,178 @@ public function downloadAsStream(array $options = [])
+            'responseType' => null,
+            'keyFile' => null,
+            'keyFilePath' => null
+        ];
+
+        $seconds = $expires->get()->format('U');
+        if ($seconds < time()) {
+            throw new \InvalidArgumentException('Expiration cannot be in the past.');
+        }
+
+        $allowedMethods = ['GET', 'PUT', 'DELETE'];
+        $options['method'] = strtoupper($options['method']);
+        if (!in_array($options['method'], $allowedMethods)) {
+            throw new \InvalidArgumentException('$options.method must be one of `GET`, `PUT` or `DELETE`.');
+        }
+
+        if ($options['keyFilePath']) {
+            if (!file_exists($options['keyFilePath'])) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Keyfile path %s does not exist.',
+                    $options['keyFilePath']
+                ));
+            }
+
+            $keyFile = json_decode(file_get_contents($options['keyFilePath']), true);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                throw new \InvalidArgumentException(sprintf(
+                    'Keyfile path %s does not contain valid json.',
+                    $options['keyFilePath']
+                ));
+            }
+        } elseif ($options['keyFile']) {
+            $keyFile = $options['keyFile'];
+        } else {
+            $requestWrapper = $this->connection->requestWrapper();
+            $keyFile = $requestWrapper->keyFile();
+        }
+
+        if (!isset($keyFile['private_key']) || !isset($keyFile['client_email'])) {
+            throw new \RuntimeException(
+                'Keyfile does not provide required information. ' .
+                'Please ensure keyfile includes `private_key` and `client_email`.'
+            );
+        }
+
+        $options['cname'] = trim($options['cname'], '/');
+
+        $resource = sprintf('/%s/%s', $this->identity['bucket'], $this->identity['object']);
+
+        $headers = [];
+        foreach ($options['headers'] as $name => $value) {
+            $value = (is_array($value))
+                ? implode(',', $value)
+                : $value;
+
+            $headers[] = $name .':'. $value;
+        }
+
+        if ($headers) {
+            $headers[] = "";
+        }
+
+        $toSign = [
+            $options['method'],
+            $options['contentMd5'],
+            $options['contentType'],
+            $seconds,
+            implode("\n", $headers) . $resource,
+        ];
+
+        $string = implode("\n", $toSign);
+        $signature = $this->signString($keyFile['private_key'], $string);
+        $encodedSignature = urlencode(base64_encode($signature));
+
+        $query = [];
+        $query[] = 'GoogleAccessId='. $keyFile['client_email'];
+        $query[] = 'Expires='. $seconds;
+        $query[] = 'Signature='. $encodedSignature;
+
+        if ($options['contentType']) {
+            $query[] = 'response-content-type='. urlencode($options['contentType']);
```

Comment:
Small nitpick: Could we have leading and following spaces whenever concatenating strings?